### PR TITLE
introduce performance profiling doc

### PR DESCRIPTION
**Pull request by awgreene**

### CODE REVIEW

File: content/en/docs/Tasks/performance-profiling.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+---
+
+## Prerequisites 
+
+- [go](https://golang.org/dl/)
+
+## Background
+
+OLM utilizes the [pprof package](https://golang.org/pkg/net/http/pprof/) from the standard go library to expose performance profiles for the OLM Operator, the Catalog Operator, and Registry Servers. Due to the sensitive nature of this data, client requests against the pprof endpoint are rejected unless they are made with the certificate data kept in the `pprof-cert secret` in the `olm namespace`. 
+Kubernetes does not provide a native way to prevent pods on cluster from iterating over the list of available ports and retrieving the data exposed. Without authetnicating the requests, OLM could leak customer usage statistics on multitenant clusters. If the aforementioned secret does not exist the pprof data will not be accessable.
```
Comment (timflannagan) [inline content/en/docs/Tasks/performance-profiling.md:15]: Just a quick driveby as I see this is still in a draft:

authetnicating -> authenticating

File: content/en/docs/Tasks/performance-profiling.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    # Replace port 8080 with 8443
+    {"op": "replace", "path": "/spec/template/spec/containers/0/ports/0", "value":{"containerPort": 8443}},
+    {"op": "replace", "path": "/spec/template/spec/containers/0/livenessProbe/httpGet/port", "value":8443},
+    {"op": "replace", "path": "/spec/template/spec/containers/0/readinessProbe/httpGet/port", "value":8443},
+
+    # Update livenessProbe and readinessProbe to use HTTPS
+    {"op": "replace", "path": "/spec/template/spec/containers/0/readinessProbe/httpGet/scheme", "value":"HTTPS"},
+    {"op": "replace", "path": "/spec/template/spec/containers/0/livenessProbe/httpGet/scheme", "value":"HTTPS"},
+]'
+deployment.apps/olm-operator patched
```
Comment (awgreene) [inline content/en/docs/Tasks/performance-profiling.md:80]: Alternatively, we could generate the manifests that include these changes as part of the OLM release.

File: content/en/docs/Tasks/performance-profiling.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+```bash
+$ export PRIVATE_KEY_FILENAME=private.key # Replace with the name of the file that contains the private key you generated.
+$ export PUBLIC_KEY_FILENAME=certificate.key # Replace with the name of the file that contains the public key you generated.
+
+$ cat << EOF | kubectl apply -f -
+apiVersion: v1
+kind: Secret
+metadata:
+  name: olm-serving-secret
```
Comment (awgreene) [inline content/en/docs/Tasks/performance-profiling.md:36]:
```suggestion
  name: olm-serving-cert
```

File: content/en/docs/Tasks/performance-profiling.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
+---
+title: "Performance Profiling Metrics"
+weight: 10
+description: >
+  The goal of this document is to familiarize you with the steps to enable and review OLM's performance profiling metrics.
```
Comment (benluddy) [inline content/en/docs/Tasks/performance-profiling.md:5]:
```suggestion
  The goal of this document is to familiarize you with the steps to enable and review OLM's performance profiling instrumentation.
```

File: content/en/docs/Tasks/performance-profiling.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+---
+
+## Prerequisites 
+
+- [go](https://golang.org/dl/)
+
+## Background
+
+OLM utilizes the [pprof package](https://golang.org/pkg/net/http/pprof/) from the standard go library to expose performance profiles for the OLM Operator, the Catalog Operator, and Registry Servers. Due to the sensitive nature of this data, OLM must be configured to use TLS Certificates before performance profiling can be enabled.
```
Comment (benluddy) [inline content/en/docs/Tasks/performance-profiling.md:14]: > and Registry Servers

Not registry-server today.

> must be configured to use TLS Certificates before performance profiling can be enabled

For clarity, since the profiling endpoints are always enabled now, this could probably be combined with the following paragraph and say something like "due to the sensitive nature of profiling data, the profiling endpoints will reject any clients that do not present a verifiable certificate. Both operators must be configured with a serving certificate and a client CA bundle in order to access the profiling endpoints".

File: content/en/docs/Tasks/performance-profiling.md [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+Once you have generated a private and public key, this data should be stored in a `TLS Secret`:
+
+```bash
+
+$ export TLS_SECRET=my-tls-secret
+$ export PRIVATE_KEY_FILENAME=private.key # Replace with the name of the file that contains the private key you generated.
+$ export PUBLIC_KEY_FILENAME=certificate.crt # Replace with the name of the file that contains the public key you generated.
+
+$ cat << EOF | kubectl apply -f -
```
Comment (benluddy) [inline content/en/docs/Tasks/performance-profiling.md:34]: Users [can also do](https://kubernetes.io/docs/concepts/configuration/secret/#tls-secrets):

```
kubectl -n my-namespace create secret tls my-name --cert=certfile --key=keyfile
```

File: content/en/docs/Tasks/performance-profiling.md [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+  name: $TLS_SECRET
+  namespace: olm
+type: kubernetes.io/tls
+data:
+  tls.key: $(base64 -w0 $PRIVATE_KEY_FILENAME)
+  tls.crt: $(base64 -w0 $PUBLIC_KEY_FILENAME)
+EOF
+```
+
+### Updating OLM to Use the TLS Secret
```
Comment (benluddy) [inline content/en/docs/Tasks/performance-profiling.md:47]: I know the chart is sort of an internal detail of the release process, but since the audience for this doc is skewed towards developers, maybe it'd be worth putting some commented-out-with-explanation lines into the values.yaml and mentioning them here as well?

File: content/en/docs/Tasks/performance-profiling.md [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+$ export DEPLOYMENT_NAME=olm-operator
+$ kubectl port-forward deployment/$DEPLOYMENT_NAME 8443:8443 -n olm
+```
+
+You can then curl the OLM `/debug/pprof` endpoint to retrieve default pprof profiles like so:
+
+```bash
+$ curl https://localhost:8443/debug/pprof/heap --cert certificate.crt --key private.key  --insecure -o olm-heap
+
+$ go tool pprof --top olm-heap
```
Comment (benluddy) [inline content/en/docs/Tasks/performance-profiling.md:105]: This could be a good place to drop a link to something like https://blog.golang.org/pprof to get readers started with how to make sense of the profiles.

File: content/en/docs/Tasks/performance-profiling.md [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+  tls.crt: $(base64 -w0 $PUBLIC_KEY_FILENAME)
+EOF
+```
+
+### Updating OLM to Use the TLS Secret
+
+Patch the OLM Deployment's pod template to use the generated TLS secret:
+
+- Defining a volume and volumeMount
+- Adding the `client-ca`, `tls-key` and `tls-crt` arguments
```
Comment (benluddy) [inline content/en/docs/Tasks/performance-profiling.md:52]:
```suggestion
- Adding the `client-ca`, `tls-key` and `tls-cert` arguments
```

File: content/en/docs/Tasks/performance-profiling.md [hunk 10]
```diff
@@ -0,0 +1,106 @@
+---
+title: "Performance Profiling Metrics"
```
Comment (benluddy) [inline content/en/docs/Tasks/performance-profiling.md:2]: I can imagine readers finding this page while trying to understand the performance of their own operator versus the OLM operators. Making the title more specific might save them some confusion or at least a click.

File: content/en/docs/Tasks/performance-profiling.md [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+Requests against the performance profiling endpoint will be rejected unless the client certificate is validated by OLM. Unfortunately, Kubernetes does not provide a native way to prevent pods on cluster from iterating over the list of available ports and retrieving the data exposed. Without authenticating the requests, OLM could leak customer usage statistics on multitenant clusters.
+
+This document will dive into the steps to [enable olm performance profiling](enable-performance-profiling) and retrieving pprof data from each component.
+
+## Enabling Performance Profiling
+
+### Creating a Certificate
+
+A valid server certiciate must be created for each component before Performance Profiling can be enabled. If you are unfamiliar with certificate generation, I recomend using the [OpenSSL](https://www.openssl.org/) tool-kit and refer to the [request certificate](https://www.openssl.org/docs/manmaster/man1/openssl-req.html) documentation.
```
Comment (timflannagan) [inline content/en/docs/Tasks/performance-profiling.md:24]:
```suggestion
A valid server certificate must be created for each component before the Performance Profiling functionality can be enabled. If you are unfamiliar with certificate generation, it's recommend to use the [OpenSSL](https://www.openssl.org/) tool-kit and refer to the [request certificate](https://www.openssl.org/docs/man1.0.2/man1/openssl-req.html) documentation.
```

File: content/en/docs/Tasks/performance-profiling.md [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+- [go](https://golang.org/dl/)
+
+## Background
+
+OLM utilizes the [pprof package](https://golang.org/pkg/net/http/pprof/) from the standard go library to expose performance profiles for the OLM Operator, the Catalog Operator, and Registry Servers. Due to the sensitive nature of this data, OLM must be configured to use TLS Certificates before performance profiling can be enabled.
+
+Requests against the performance profiling endpoint will be rejected unless the client certificate is validated by OLM. Unfortunately, Kubernetes does not provide a native way to prevent pods on cluster from iterating over the list of available ports and retrieving the data exposed. Without authenticating the requests, OLM could leak customer usage statistics on multitenant clusters.
+
+This document will dive into the steps to [enable olm performance profiling](enable-performance-profiling) and retrieving pprof data from each component.
```
Comment (timflannagan) [inline content/en/docs/Tasks/performance-profiling.md:18]: It looks like a broken link to the next header section.

```suggestion
This document will dive into the steps to [enable olm performance profiling](#enabling-performance-profiling) and retrieving pprof data from each component.
```

File: content/en/docs/Tasks/performance-profiling.md [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+```
+
+### Updating OLM to Use the TLS Secret
+
+Patch the OLM Deployment's pod template to use the generated TLS secret:
+
+- Defining a volume and volumeMount
+- Adding the `client-ca`, `tls-key` and `tls-crt` arguments
+- Replacing all mentions of port `8080` with `8443`
+- Updating the `livenessProbe` and `readinessProbe` to use HTTPS as the scheme.
```
Comment (timflannagan) [inline content/en/docs/Tasks/performance-profiling.md:54]:
```suggestion
- Updating the `livenessProbe` and `readinessProbe` to use HTTPS as the scheme
```

File: content/en/docs/Tasks/performance-profiling.md [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+
+This can by generating patching an existing OLM deployment:
```
Comment (timflannagan) [inline content/en/docs/Tasks/performance-profiling.md:56]:
```suggestion
This can by generated by patching an existing OLM deployment:
```

**Comment by awgreene** [comment]
> @awgreene this doc looks more like it belongs to the advanced task section to me. Otherwise lgtm

Agreed @anik120 - moved the document to the advanced task directory.

File: content/en/docs/advanced-tasks/performance-profiling.md [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+
+Once you have generated a private and public key, this data should be stored in a `TLS Secret`:
+
+```bash
+
+$ export TLS_SECRET=my-tls-secret
+$ export PRIVATE_KEY_FILENAME=private.key # Replace with the name of the file that contains the private key you generated.
+$ export PUBLIC_KEY_FILENAME=certificate.crt # Replace with the name of the file that contains the public key you generated.
+
+$ kubectl -n my-namespace create secret tls my-name --cert=$PUBLIC_KEY_FILENAME --key=$PRIVATE_KEY_FILENAME
```
Comment (estroz) [inline content/en/docs/advanced-tasks/performance-profiling.md:34]: Missing codeblock end.
````suggestion
$ kubectl -n my-namespace create secret tls my-name --cert=$PUBLIC_KEY_FILENAME --key=$PRIVATE_KEY_FILENAME
```
````

File: content/en/docs/advanced-tasks/performance-profiling.md [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+
+### Creating a Certificate
+
+A valid server certificate must be created for each component before the Performance Profiling functionality can be enabled. If you are unfamiliar with certificate generation, we recommend using the [OpenSSL](https://www.openssl.org/) tool-kit and refer to the [request certificate](https://www.openssl.org/docs/man1.1.1/man1/openssl-req.html) documentation.
+
+Once you have generated a private and public key, this data should be stored in a `TLS Secret`:
+
+```bash
+
+$ export TLS_SECRET=my-tls-secret
```
Comment (estroz) [inline content/en/docs/advanced-tasks/performance-profiling.md:30]: This variable isn't used anywhere in this codeblock. Is it supposed to be the secret name?

**Comment by anik120** [comment]
Closing due to inactivity

**Comment by anik120** [comment]
Closed this by mistake oops.

**Review by perdasilva — APPROVED** [review]
